### PR TITLE
Read Adapters (Get & Scan) changes

### PR DESCRIPTION
**Pull request by rahulKQL**
## This PR contains

 - Updated `ReadHooks` to adapt to `v2.models.Query`
 - Updated `HBaseRequestAdapter`'s adapt to `Get` & `Scan`.
 - `ScanAdapter` is updated find range on the basis of Guvava's `RangeSet`.
 - `RowRangeAdapter` - Updated this class to process `ByteStringRange` privosuly processing `RowSet`.
- Updated `TemplateUtil.RequestValueProvider` constructor to avail necessary properties for Query and RequestContext instantiation.
 - Updates in `com.google.cloud.bigtable.beam.CloudBigtableScanConfiguration` & `com.google.cloud.bigtable.dataflow.CloudBigtableScanConfiguration`.
 - Other files have minor changes (mostly addition of `RequestContext` for compatibility.

### CODE REVIEW

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BatchExecutor.java [hunk 1]
```diff
@@ -201,7 +208,9 @@ public BatchExecutor(BigtableSession session, HBaseRequestAdapter requestAdapter
   private ListenableFuture<?> issueAsyncRequest(BulkOperation bulkOperation, Row row) {
     try {
       if (row instanceof Get) {
-        return bulkOperation.bulkRead.add(requestAdapter.adapt((Get) row));
+//        Query request = requestAdapter.adapt((Get) row);
```
Comment (sduskis) [inline bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BatchExecutor.java]: Please cleanup any commented code.

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/adapters/HBaseRequestAdapter.java [hunk 2]
```diff
@@ -16,12 +16,9 @@
...
 
 import com.google.api.core.InternalApi;
+import com.google.bigtable.v2.ReadRowsRequest;
 import com.google.cloud.bigtable.data.v2.internal.RequestContext;
-import com.google.cloud.bigtable.data.v2.models.InstanceName;
-import com.google.cloud.bigtable.data.v2.models.Mutation;
-import com.google.cloud.bigtable.data.v2.models.MutationApi;
-import com.google.cloud.bigtable.data.v2.models.ReadModifyWriteRow;
-import com.google.cloud.bigtable.data.v2.models.RowMutation;
+import com.google.cloud.bigtable.data.v2.models.*;
```
Comment (sduskis) [inline bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/adapters/HBaseRequestAdapter.java]: Please change your settings to not use `*`.

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/adapters/read/ReadHooks.java [hunk 3]
```diff
@@ -15,11 +15,11 @@
  */
 package com.google.cloud.bigtable.hbase.adapters.read;
 
-import com.google.bigtable.v2.ReadRowsRequest;
+import com.google.cloud.bigtable.data.v2.models.Query;
 import com.google.common.base.Function;
 
 /**
- * Hooks for modifying a ReadRowsRequest before being sent to Cloud Bigtable.
+ * Hooks for modifying a Query before being sent to Cloud Bigtable.
```
Comment (sduskis) [inline bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/adapters/read/ReadHooks.java]: Please use `{@link ...}` here, even though we didn't use it before.

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/adapters/read/ReadHooks.java [hunk 4]
```diff
@@ -30,17 +30,17 @@
 public interface ReadHooks {
 
   /**
-   * Add a Function that will modify the ReadRowsRequest before it is sent to Cloud Bigtable.
+   * Add a Function that will modify the Query before it is sent to Cloud Bigtable.
```
Comment (sduskis) [inline bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/adapters/read/ReadHooks.java]: `{@link }` please.

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/adapters/read/RowRangeAdapter.java [hunk 5]
```diff
@@ -67,7 +69,7 @@
         break;
       default:
         throw new IllegalArgumentException("Unexpected start key case: " +
-            btRange.getStartKeyCase());
+                btRange.getStartKeyCase());
```
Comment (sduskis) [inline bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/adapters/read/RowRangeAdapter.java]: Please fix your spacing to be similar to the current style.  There should be an eclipse style xml file at the root of the project.

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/adapters/read/RowRangeAdapter.java [hunk 6]
```diff
@@ -92,6 +94,17 @@
       default:
         throw new IllegalArgumentException("Unexpected end key case: " + btRange.getEndKeyCase());
     }
+
+    return boundedRange(startBound, startKey, endBound, endKey);
+  }
+
+  Range<RowKeyWrapper> boundedRange(BoundType startBound, ByteString startKey,
+          BoundType endBound, ByteString endKey){
+    // Bigtable doesn't allow empty row keys, so an empty start row which is open or closed is
```
Comment (sduskis) [inline bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/adapters/read/RowRangeAdapter.java]: Please add quotes (`"`) around the empty string for clarity.

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/adapters/read/RowRangeAdapter.java [hunk 7]
```diff
@@ -105,35 +118,34 @@
       return Range.downTo(new RowKeyWrapper(startKey), startBound);
     } else {
       return Range.range(new RowKeyWrapper(startKey), startBound,
-          new RowKeyWrapper(endKey), endBound);
+              new RowKeyWrapper(endKey), endBound);
```
Comment (sduskis) [inline bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/adapters/read/RowRangeAdapter.java]: Spacing again...

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/adapters/read/RowRangeAdapter.java [hunk 8]
```diff
@@ -105,35 +118,34 @@
...
     }
   }
 
   /**
-   * Convert guava's RangeSet to Bigtable's RowSet. Please note that this will convert
+   * Convert guava's RangeSet to Bigtable's ByteStringRange. Please note that this will convert
```
Comment (sduskis) [inline bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/adapters/read/RowRangeAdapter.java]: `{@link}`

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/adapters/read/ScanAdapter.java [hunk 9]
```diff
@@ -137,60 +140,45 @@ public void throwIfUnsupportedScan(Scan scan) {
...
-          if (!OPEN_CLOSED_AVAILABLE || !scan.includeStopRow()) {
-            // the default for stop is open
-            range.setEndKeyOpen(stopRow);
-          } else {
-            range.setEndKeyClosed(stopRow);
-          }
-        }
-        rowSetBuilder.addRowRanges(range);
+        rangeSet.add(rowRangeAdapter
+                .boundedRange(startBound, startRow, endBound,stopRow));
```
Comment (sduskis) [inline bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/adapters/read/ScanAdapter.java]: Please add a space after the `,`

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BatchExecutor.java (hunk 1 above)
Comment (rahulKQL) [reply bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BatchExecutor.java]: Thanks, will remove these commented lines.

File: bigtable-client-core-parent/bigtable-hbase/src/test/java/com/google/cloud/bigtable/hbase/TestBatchExecutor.java [hunk 10]
```diff
@@ -132,9 +132,11 @@ public void describeMismatch(Object item, Description mismatchDescription) {
   private ListenableFuture mockFuture;
 
   private HBaseRequestAdapter requestAdapter;
+
+  private BigtableOptions options;
```
Comment (sduskis) [inline bigtable-client-core-parent/bigtable-hbase/src/test/java/com/google/cloud/bigtable/hbase/TestBatchExecutor.java]: Please add a new line after the member variables.

File: bigtable-client-core-parent/bigtable-hbase/src/test/java/com/google/cloud/bigtable/hbase/TestBatchExecutor.java [hunk 11]
```diff
@@ -238,7 +240,11 @@ public void testPartialResults() throws Exception {
     Object[] results = new Object[2];
 
     try {
-      createExecutor(BigtableOptions.getDefaultOptions()).batch(gets, results);
+      final BigtableOptions options = BigtableOptions.builder()
```
Comment (sduskis) [inline bigtable-client-core-parent/bigtable-hbase/src/test/java/com/google/cloud/bigtable/hbase/TestBatchExecutor.java]: Do we need this variable, given the member variable you added?

File: bigtable-client-core-parent/bigtable-hbase/src/test/java/com/google/cloud/bigtable/hbase/adapters/filters/TestPageFilterAdapter.java [hunk 12]
```diff
@@ -110,14 +113,18 @@ public void mustPassAllIsNotSupportedBelowTopLevel() {
...
     PageFilter pageFilter = new PageFilter(20);
     Filters.Filter adaptedFilter = pageFilterAdapter.adapt(context, pageFilter);
     Assert.assertNull("PageFilterAdapter should not return a Filters.Filter.", adaptedFilter);
 
-    ReadRowsRequest request = ReadRowsRequest.newBuilder().setRowsLimit(100).build();
-    ReadRowsRequest postHookRequest = hooks.applyPreSendHook(request);
-    Assert.assertEquals(20, postHookRequest.getRowsLimit());
+    Query query = Query.create(TABLE_ID).limit(100);
+    Query postHookRequest = hooks.applyPreSendHook(query);
+    Assert.assertEquals(20, postHookRequest.toProto(requestContext).getRowsLimit());
```
Comment (sduskis) [inline bigtable-client-core-parent/bigtable-hbase/src/test/java/com/google/cloud/bigtable/hbase/adapters/filters/TestPageFilterAdapter.java]: Can we get the rows limit directly from the Query?

File: bigtable-client-core-parent/bigtable-hbase/src/test/java/com/google/cloud/bigtable/hbase/adapters/read/TestGetAdapter.java [hunk 13]
```diff
@@ -68,60 +77,106 @@ private Get makeValidGet(byte[] rowKey) throws IOException {
...
-    ByteString adaptedRowKey = rowRequestBuilder.getRows().getRowKeys(0);
+    Get get = makeValidGet(dataHelper.randomData(PREFIX_DATA));
+    Query query = Query.create(TABLE_ID);
+    getAdapter.adapt(get, throwingReadHooks, query);
+
+    ByteString adaptedRowKey = query.toProto(requestContext).getRows().getRowKeys(0);
     Assert.assertEquals(
-        new String(get.getRow(), StandardCharsets.UTF_8),
-        adaptedRowKey.toStringUtf8());
+            new String(get.getRow(), StandardCharsets.UTF_8),
```
Comment (sduskis) [inline bigtable-client-core-parent/bigtable-hbase/src/test/java/com/google/cloud/bigtable/hbase/adapters/read/TestGetAdapter.java]: Spacing again...

File: bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/TemplateUtils.java [hunk 14]
```diff
@@ -99,8 +106,9 @@ public ReadRowsRequest get() {
         }
 
         ReadHooks readHooks = new DefaultReadHooks();
-        ReadRowsRequest.Builder builder = Adapters.SCAN_ADAPTER.adapt(scan, readHooks);
-        cachedRequest = readHooks.applyPreSendHook(builder.build());
+        Query query = Query.create(options.getBigtableTableId().get());
+        Adapters.SCAN_ADAPTER.adapt(scan, readHooks, query);
+        cachedRequest = ((Query)readHooks.applyPreSendHook(query)).toProto(requestContext);
```
Comment (sduskis) [inline bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/TemplateUtils.java]: Why do you need to cast `readHooks .applyPreSendHook` to a `Query`?

File: bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/TemplateUtils.java [hunk 15]
```diff
@@ -65,16 +69,19 @@ public static CloudBigtableTableConfiguration BuildImportConfig(ImportOptions op
...
-    RequestValueProvider(
-        ValueProvider<String> start,
-        ValueProvider<String> stop,
-        ValueProvider<Integer> maxVersion,
-        ValueProvider<String> filter) {
-      this.start = start;
-      this.stop = stop;
-      this.maxVersion = maxVersion;
-      this.filter = filter;
+    RequestValueProvider(ExportOptions options) {
```
Comment (sduskis) [inline bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/TemplateUtils.java]: This should definitely be a different PR.

File: bigtable-client-core-parent/bigtable-hbase/src/test/java/com/google/cloud/bigtable/hbase/adapters/read/TestScanAdapter.java [hunk 16]
```diff
@@ -307,7 +338,26 @@ public void testEmptyFilterList(){
...
     Scan scan = new Scan().setLimit(10);
-    int adaptedLimit = (int) scanAdapter.adapt(scan, throwingReadHooks).build().getRowsLimit();
+    Query query = Query.create(TABLE_ID);
+    scanAdapter.adapt(scan, throwingReadHooks, query);
+    int adaptedLimit = (int) query.toProto(requestContext).getRowsLimit();
     Assert.assertEquals(scan.getLimit(), adaptedLimit);
   }
+
+  @Test
+  public void testIsGetScan() throws IOException{
```
Comment (sduskis) [inline bigtable-client-core-parent/bigtable-hbase/src/test/java/com/google/cloud/bigtable/hbase/adapters/read/TestScanAdapter.java]: This is a good candidate for a separate PR.

File: bigtable-client-core-parent/bigtable-hbase/src/test/java/com/google/cloud/bigtable/hbase/adapters/read/TestScanAdapter.java [hunk 17]
```diff
@@ -55,17 +57,24 @@
 @RunWith(JUnit4.class)
 public class TestScanAdapter {
 
+  public static final String STOP_KEY = "stopKey";
```
Comment (sduskis) [inline bigtable-client-core-parent/bigtable-hbase/src/test/java/com/google/cloud/bigtable/hbase/adapters/read/TestScanAdapter.java]: All cosmetic changes like static variables should be introduced in a separate PR.

File: bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/TemplateUtils.java (hunk 14 above)
Comment (rahulKQL) [reply bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/TemplateUtils.java]: It was showing up as an error in IDE, so added, must be because I hadn't run maven while making reference change as it is dependent on shaded jars. Would remove the unnecessary Cast.

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/adapters/filters/PageFilterAdapter.java [hunk 18]
```diff
@@ -43,13 +44,12 @@
...
-    context.getReadHooks().composePreSendHook(new Function<ReadRowsRequest, ReadRowsRequest>() {
+    context.getReadHooks().composePreSendHook(new Function<Query, Query>() {
       @Override
-      public ReadRowsRequest apply(ReadRowsRequest request) {
-        return request.toBuilder().setRowsLimit(pageSize).build();
+      public Query apply(Query query) {
+        return query.limit(pageSize);
       }
     });
-    // This filter cannot be translated to a RowFilter, all logic is done as a read hook.
```
Comment (sduskis) [inline bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/adapters/filters/PageFilterAdapter.java]: Please keep the comment.

File: bigtable-dataflow-parent/bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/dataflow/CloudBigtableScanConfiguration.java [hunk 19]
```diff
@@ -177,8 +180,12 @@ public CloudBigtableScanConfiguration build() {
         if (scan == null) {
           scan = new Scan();
         }
-        ReadRowsRequest.Builder builder = Adapters.SCAN_ADAPTER.adapt(scan, readHooks);
-        request = readHooks.applyPreSendHook(builder.build());
+        Query query = Query.create(tableId);
+        Adapters.SCAN_ADAPTER.adapt(scan, readHooks, query);
+        RequestContext requestContext =
+            RequestContext.create(InstanceName.of(projectId, instanceId), "");
+        readHooks.applyPreSendHook(query);
```
Comment (rahulKQL) [inline bigtable-dataflow-parent/bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/dataflow/CloudBigtableScanConfiguration.java]: `appProfileId` is not available at this class... I added an empty string to for the same in `RequestContext.create()` constructor. should this be fine?

File: bigtable-client-core-parent/bigtable-hbase/src/test/java/com/google/cloud/bigtable/hbase/adapters/read/TestRowRangeAdapter.java [hunk 20]
```diff
@@ -45,7 +52,9 @@ public void testEmptyRowSet() {
     RangeSet<RowKeyWrapper> out = adapter.rowSetToRangeSet(in);
 
     assertEquals(ImmutableRangeSet.builder().build(), out);
-    assertEquals(in, adapter.rangeSetToRowSet(out));
+    Query query = Query.create(TABLE_ID);
```
Comment (sduskis) [inline bigtable-client-core-parent/bigtable-hbase/src/test/java/com/google/cloud/bigtable/hbase/adapters/read/TestRowRangeAdapter.java]: This is repeated over and over.  Please refactor this logic into a new method

File: bigtable-client-core-parent/bigtable-hbase/src/test/java/com/google/cloud/bigtable/hbase/adapters/read/TestScanAdapter.java (hunk 17 above)
Comment (sduskis) [reply bigtable-client-core-parent/bigtable-hbase/src/test/java/com/google/cloud/bigtable/hbase/adapters/read/TestScanAdapter.java]: @rahulKQL, please create a new PR for the new variables.

File: bigtable-dataflow-parent/bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/dataflow/CloudBigtableScanConfiguration.java (hunk 19 above)
Comment (sduskis) [reply bigtable-dataflow-parent/bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/dataflow/CloudBigtableScanConfiguration.java]: You should be able to get the appProfileId from the `additionalConfiguration`.

Comment (rahulKQL) [reply bigtable-dataflow-parent/bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/dataflow/CloudBigtableScanConfiguration.java]: The reason I thought `additionConfiguration` might not have `appProfileId` because `dataflow.CloudBigtableScanConfiguration` doesn't have  **`withAppProfileId()`**, but it's present  in `beam.CloudBigtableScanConfiguration`.

Would it be available through `withConfiguration(key, value)`?

File: bigtable-client-core-parent/bigtable-hbase/src/test/java/com/google/cloud/bigtable/hbase/adapters/read/TestScanAdapter.java (hunk 17 above)
Comment (rahulKQL) [reply bigtable-client-core-parent/bigtable-hbase/src/test/java/com/google/cloud/bigtable/hbase/adapters/read/TestScanAdapter.java]: I would revert these to their previous line, I just refactor these two constants to group with rest of the constant. I added these constant in #2013

File: bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/CloudBigtableScanConfiguration.java [hunk 21]
```diff
@@ -276,11 +280,21 @@ public CloudBigtableScanConfiguration build() {
         if (scan == null) {
           scan = new Scan();
         }
-        ReadRowsRequest.Builder builder = Adapters.SCAN_ADAPTER.adapt(scan, readHooks);
-        request =  StaticValueProvider.of(readHooks.applyPreSendHook(builder.build()));
+        Query query = Query.create(tableId.get());
+        Adapters.SCAN_ADAPTER.adapt(scan, readHooks, query);
+        ValueProvider<String> appProfileIdValue =
```
[the remark this answers is not in the document]
Comment (rahulKQL) [reply bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/CloudBigtableScanConfiguration.java]: @igorbernstein2, This is where we have to use tableId & other value providers, so working on refactoring these

**Comment by rahulKQL** [comment]
Dependent on #2025  as projectId, instanceId & tableId would not in available in RequestValueProvider#get untill later part of execution.

File: bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/CloudBigtableScanConfiguration.java [hunk 22]
```diff
@@ -276,11 +285,33 @@ public CloudBigtableScanConfiguration build() {
...
+        }
+
+        Query query = Query.create(PLACEHOLDER_TABLE_ID);
+        Adapters.SCAN_ADAPTER.adapt(scan, readHooks, query);
+        readHooks.applyPreSendHook(query);
+
+        RequestContext requestContext = RequestContext
+            .create(materializedProjectId, materializedInstanceId, materializedAppProfileId);
+
+        request = StaticValueProvider.of(query.toProto(requestContext));
```
Comment (rahulKQL) [inline bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/CloudBigtableScanConfiguration.java]: Hi @sduskis / @igorbernstein2 

Are there any chances of ProjectId & instanceId available at this place...(_assuming there could be that's why checking ValueProviders for projectId, instanceId & appProfileId_)?

Also recalculating ReadRowRequest at [Line#351](https://github.com/GoogleCloudPlatform/cloud-bigtable-client/pull/2011/files#diff-5836fc24866e6a80ee2b663eb55592d4R351).

Does this look good to proceed?

File: bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/TemplateUtils.java [hunk 23]
```diff
@@ -95,8 +109,12 @@ public ReadRowsRequest get() {
...
 
         ReadHooks readHooks = new DefaultReadHooks();
-        ReadRowsRequest.Builder builder = Adapters.SCAN_ADAPTER.adapt(scan, readHooks);
-        cachedRequest = readHooks.applyPreSendHook(builder.build());
+        Query query = Query.create(PLACEHOLDER_TABLE_ID);
+        RequestContext requestContext = RequestContext
+            .create(PLACEHOLDER_PROJECT_ID, PLACEHOLDER_INSTANCE_ID, PLACEHOLDER_APP_PROFILE_ID);
+        readHooks.applyPreSendHook(query);
+
+        cachedRequest = query.toProto(requestContext);
```
Comment (rahulKQL) [inline bigtable-dataflow-parent/bigtable-beam-import/src/main/java/com/google/cloud/bigtable/beam/TemplateUtils.java]: Can we use placeholders for all four the values?

Then override it in [RequestWithTableNameValueProvider#get](https://github.com/GoogleCloudPlatform/cloud-bigtable-client/pull/2011/files#diff-5836fc24866e6a80ee2b663eb55592d4R347) with actual tableName & appProfileId.

File: bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/CloudBigtableScanConfiguration.java [hunk 24]
```diff
@@ -276,11 +285,33 @@ public CloudBigtableScanConfiguration build() {
         if (scan == null) {
           scan = new Scan();
         }
-        ReadRowsRequest.Builder builder = Adapters.SCAN_ADAPTER.adapt(scan, readHooks);
-        request =  StaticValueProvider.of(readHooks.applyPreSendHook(builder.build()));
+        String materializedProjectId = PLACEHOLDER_PROJECT_ID;
+        String materializedInstanceId = PLACEHOLDER_INSTANCE_ID;
+        String materializedAppProfileId = PLACEHOLDER_APP_PROFILE_ID;
+
+        if(projectId != null && Strings.isNullOrEmpty(projectId.get())){
```
Comment (sduskis) [inline bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/CloudBigtableScanConfiguration.java]: Please don't call `.get()` here.  Please assume that projectId & etc are not available.  Please see `CloudBigtableConfiguration.getDisplayData()` for proper checking of value provider availability.

Comment (rahulKQL) [reply bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/CloudBigtableScanConfiguration.java]: Updated code, Now adapting through place holder for all four values i.e. projectId, instanceId, appProfileId & TableId.

**Comment by rahulKQL** [comment]
**Added this change in last commit:**
- Removed `projectId.get()` & `instanceId.get()` from `CloudBigtableScanConfiguration#build` method.
- Now checking for availability of ValueProvider using the `isAccessible` method.

File: bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/CloudBigtableScanConfiguration.java (hunk 22 above)
Comment (sduskis) [reply bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/CloudBigtableScanConfiguration.java]: We cannot proceed.  The `build()` will not necessarily have the available values in place.  Please always wrap `ValueProvider.get()` in a call to `ValueProvider.isAccessible()`

File: bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/CloudBigtableScanConfiguration.java [hunk 25]
```diff
@@ -297,26 +328,41 @@ public CloudBigtableScanConfiguration build() {
...
     private ReadRowsRequest cachedRequest;
+    private Map<String, ValueProvider<String>> additionalConfiguration;
 
     RequestWithTableNameValueProvider(
         ValueProvider<String> projectId,
         ValueProvider<String> instanceId,
         ValueProvider<String> tableId,
-        ValueProvider<ReadRowsRequest> request) {
+        ValueProvider<ReadRowsRequest> request,
+        Map<String, ValueProvider<String>> additionalConfiguration) {
```
Comment (sduskis) [inline bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/CloudBigtableScanConfiguration.java]: Please add `ValueProvider<String> appProfileId` above request, and remove the `additionalConfiguration`

File: bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/CloudBigtableScanConfiguration.java (hunk 22 above)
Comment (rahulKQL) [reply bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/CloudBigtableScanConfiguration.java]: Instead of using `Query.create(PLACEHOLDER_TABLE_ID)`, 
Shall I use `ReadRowsRequest.getDefaultInstance()`, however in both the case I would have to use Placeholders for `projectId`, `instanceId` & `appProfileId` inside `build()`.
then we can check for PlaceHolders inside RequestWithTableNameValueProvider#get and replace them with actual values.

**Comment by sduskis** [comment]
@rahulKQL, I added a `CloudBigtableScanConfiguration.Builder.withQuery()` that should work for `CloudBigtableScanConfiguration`.  Please rebase, and update `CloudBigtableScanConfiguration` to use that new method, and use the same approach in `TemplateUtils`.

**Comment by rahulKQL** [comment]
- Updated [CloudBigtableScanConfiguration#withScan](https://github.com/GoogleCloudPlatform/cloud-bigtable-client/pull/2011/files#diff-5836fc24866e6a80ee2b663eb55592d4L80) to adapt the latest [ScanAdapter#adapt](https://github.com/GoogleCloudPlatform/cloud-bigtable-client/pull/2011/files#diff-a0da3e054ab30c034391e2c0e7de098cR141) with Query.

- Have updated CloudBigtableScanConfiguration#withQuery to use Placeholders to make it consistent between [TemplateUtil](https://github.com/GoogleCloudPlatform/cloud-bigtable-client/pull/2011/files#diff-7505ccb11739a5266baf3571a5b732a5) & [withQuery](https://github.com/GoogleCloudPlatform/cloud-bigtable-client/pull/2011/files#diff-5836fc24866e6a80ee2b663eb55592d4L88).

Please have a look.
